### PR TITLE
Fix/edit address api improvements

### LINKED ISSUE #9: [Breaking change] Edit address API improvements
As a developer I want to have an ease to use API of the EditAddressWidget so that I can integrate that component with a minimal effort.

ACs

- [x] All icons to be moved from the widget constructor to the theme extension
- [x] Error mappers to be grouped in one abstract class
- [x] All business logic to be consolidated in one service
- [ ] Update the documentation 
  - [ ] Add an example with the minimal configuration
  - [ ] Add an example with the full configuration
- [ ] Update the package version as well as the CHANGELOG
- [ ] Add golden tests for the not-empty cases
- [x] Add Edit address in the example project
- [x] Add onChanged function receiving the changed address

**Current API**
```dart
EditAddressWidget({
    required this.cityErrorMapper,
    required this.addressErrorMapper,
    required this.validator,
    required this.searchCountryService,
    this.editAddressService = const _DefaultEditAddressService(),
    this.addressModel = _defaultAddressModel,
    this.editAddressLocalizedStrings,
    this.type = UserProfileCardTypes.mailingAddress,
    this.configuration = const EditAddressConfiguration(),
    this.countryCustomIcon,
    this.editCountryFieldType = EditFieldType.dropdown,
    this.cityCustomIcon,
    this.editCityFieldType = EditFieldType.editfield,
    this.addressCustomIcon,
    this.editAddressFieldType = EditFieldType.editfield,
    this.editContactAddressErrorBuilder,
    this.searchCountryCustomBuilders,
    Key? key,
  })
```

**Improved API**
`edit_address_widget.dart`
```dart
EditAddressWidget({
    required this.translateError,
    required this.service,
    this.onChanged, // Function (AddressModel?)
    this.initialAddress
    this.searchCountryBuilders,
    this.editContactAddressErrorBuilder,
    this.localizedStrings,
    this.configuration = const EditAddressConfiguration(),
    Key? key,
  })
```

`edit_address_service.dart`
```dart
abstract class EditAddressService<T> implements SearchPickerService<T>, TextFieldValidator<String> {
    Future<AddressModel> saveAddress(AddressModel addressModel);

    Future<List<T>> getCountries();

    List<T> getCountryPlaceholderList();

    // Default implementation
    Future<List<T>> filteredCountriesByName(List<T> list, String? searchParam);

    // Default implementation lengthIsNotEmpty
    Future<T> validateCityOnSubmit(String text);

    // Default implementation lengthIsNotEmpty
    Future<T> validateStreetOnSubmit(String text);
    
    // Default implementation lengthIsNotEmpty
    void validateCityOnType(String text);

    // Default implementation lengthIsNotEmpty
    void validateStreetOnType(String text);
}
```

`edit_address_theme.dart`
```dart
EditAddressTheme(
    this.countryIcon,
...
)
TextFieldDialogTheme(
    this.cityIcon,
    this.addressIcon,
)
```
Note: 
For the default implementation of the methods with this check : `lengthIsNotEmpty`
In the example project we need default implementation of the city and street error mappers to map the error to RxFieldException in order to see it on the screen, when we try to save the empty value.
Also currently there is a default implementation for the error that we see when we try to SAVE. For this method: Future<AddressModel> saveAddress(AddressModel addressModel); In the example project, we should provide implementation also for it, otherwise there will be no error on the screen as it shows now.

### PR DESCRIPTION
**Pull request by StanislavPrime**
Closes #9 
This PR is work in progress

- [x] Remove countryCustomIcon, cityCustomIcon, addressCustomIcon parameters from EditAddressWidget
- [x] Remove editCountryFieldType, editCityFieldType, editAddressFieldType parameters from EditAddressWidget and hardcode the previously default values, so there is no exception if EditFieldType.custom is selected.
- [x] Replace the 2 city and street error mapper function parameters by a single `Function(Object error) translateError` function.
- [x] In `EditAddressWidget`'s parameters replace the different services: `TextFieldValidator`, `SearchPickerService`, with one service: EditAddressService, which consolidates the whole business logic API for the librart. In `EditAddressService` were added additional methods: `validateCityOnSubmit`,`validateStreetOnSubmit`,`validateCityOnType`,`validateStreetOnType`, replacing the previously used: `validateOnSubmit` and `validateOnType`, providing more granular validation on the city and address validation. Methods: `getItems()` becomes `getCountries()`, `filteredListByName()`  becomes `filteredCountries()`, `getPlaceholderList()` becomes `getCountryPlaceholderList()`
- [x]  In `EditAddressWidget` add as a new parameter function: `final Function(AddressModel? addressModel)? onChanged;`
- [x]  Add golden tests for the not-empty cases

### CODE REVIEW

File: packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart [hunk 1]
```diff
@@ -1,7 +1,80 @@
...
+      return list;
+    }
+    return list
+        .where((item) =>
+            (item.toString()).toLowerCase().contains(searchParam.toLowerCase()))
+        .toList();
+  }
+
+  Future<String> validateCityOnSubmit(String text) async {
+    await Future.delayed(const Duration(seconds: 1));
```
Comment (StanevPrime) [inline packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart:53]: I strongly suggest to not put artificial delays into the package itself.

File: packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart [hunk 2]
```diff
@@ -1,7 +1,80 @@
...
+    if (text.trim().isEmpty) {
+      await Future.delayed(const Duration(seconds: 1));
+      throw Exception('A value should be provided');
+    }
+    return text;
+  }
+
+  Future<String> validateStreetOnSubmit(String text) async {
+    if (text.trim().isEmpty) {
+      await Future.delayed(const Duration(seconds: 1));
```
Comment (StanevPrime) [inline packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart:63]: I strongly suggest to not put artificial delays into the package itself.

File: packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart [hunk 3]
```diff
@@ -1,7 +1,80 @@
...
+      throw Exception('A value should be provided');
+    }
+    return text;
+  }
+
+  void validateCityOnType(String text) {
+    if (text.trim().isEmpty) {
+      throw Exception('A value should be provided');
```
Comment (StanevPrime) [inline packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart:71]: Please use the business error models instead of throwing an exception

File: packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart [hunk 4]
```diff
@@ -1,7 +1,80 @@
...
+    }
+  }
+
+  void validateStreetOnType(String text) {
+    if (text.trim().isEmpty) {
+      throw Exception('A value should be provided');
```
Comment (StanevPrime) [inline packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart:77]: Please use the business error models instead of throwing an exception

File: packages/widget_toolkit/lib/src/lib_edit_address/ui_components/edit_address_form.dart [hunk 5]
```diff
@@ -71,7 +60,8 @@ class EditAddressForm<T extends PickerItemModel> extends StatelessWidget {
                       bloc.events.saveCountry();
                     },
                     modalConfiguration: countryPickerModalConfiguration,
-                    searchCountryService: searchCountryService,
+                    searchCountryService:
+                        CountryService(editAddressService: editAddressService),
```
Comment (StanevPrime) [inline packages/widget_toolkit/lib/src/lib_edit_address/ui_components/edit_address_form.dart:64]: Please avoid creating instances of CountryService on each rebuild

File: packages/widget_toolkit/lib/src/lib_edit_address/views/edit_address_page.dart [hunk 6]
```diff
@@ -122,87 +95,83 @@ class EditAddressPage<T extends PickerItemModel> extends StatelessWidget {
...
-                                    .read<EditAddressBlocType>()
-                                    .events
-                                    .saveAddress();
-                              } else {
-                                if (address.hasData) {
-                                  saveAddress?.call(address.data!);
+                      if (!snapshot.hasData)
+                        EditAddressForm.withDependencies<T>(
+                            context, addressModel, (addressModel) {
+                          return context
```
Comment (StanevPrime) [inline packages/widget_toolkit/lib/src/lib_edit_address/views/edit_address_page.dart:127]: For better readability you can use an expression body

**Review by StanevPrime — CHANGES_REQUESTED** [review]

File: packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart (hunk 4 above)
Comment (StanislavPrime) [reply packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart:77]: I have changed the `Exception` to `GenericErrorModel`

File: packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart (hunk 3 above)
Comment (StanislavPrime) [reply packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart:71]: I have changed the `Exception` to `GenericErrorModel`

File: packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart (hunk 2 above)
Comment (StanislavPrime) [reply packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart:63]: The delay was deleted.

File: packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart (hunk 1 above)
Comment (StanislavPrime) [reply packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart:53]: The delay was deleted.

File: packages/widget_toolkit/lib/src/lib_edit_address/views/edit_address_page.dart (hunk 6 above)
Comment (StanislavPrime) [reply packages/widget_toolkit/lib/src/lib_edit_address/views/edit_address_page.dart]: The `return` was changed to expression body here and on some other places

File: packages/widget_toolkit/lib/src/lib_edit_address/ui_components/edit_address_form.dart (hunk 5 above)
Comment (StanislavPrime) [reply packages/widget_toolkit/lib/src/lib_edit_address/ui_components/edit_address_form.dart]: I have moved the creation of the services; CityService, StreetService and CountryService in the edit_address_dependencies.dart file

File: packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart [hunk 7]
```diff
@@ -1,7 +1,77 @@
...
+    }
+    return list
+        .where((item) =>
+            (item.toString()).toLowerCase().contains(searchParam.toLowerCase()))
+        .toList();
+  }
+
+  Future<String> validateCityOnSubmit(String text) async {
+    if (text.trim().isEmpty) {
+      throw GenericErrorModel(customMessage:'A value should be provided');
```
Comment (StanevPrime) [inline packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart:54]: Using the `GenericErrorModel` makes the translation of the messaged shown to the user impossible

File: packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart [hunk 8]
```diff
@@ -1,7 +1,77 @@
...
+    }
+    return text;
+  }
+
+  Future<String> validateStreetOnSubmit(String text) async {
+    if (text.trim().isEmpty) {
+      throw GenericErrorModel(customMessage: 'A value should be provided');
```
Comment (StanevPrime) [inline packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart:61]: Using the `GenericErrorModel` makes the translation of the messaged shown to the user impossible

File: packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart [hunk 9]
```diff
@@ -1,7 +1,77 @@
...
+    }
+    return text;
+  }
+
+  void validateCityOnType(String text) {
+    if (text.trim().isEmpty) {
+      throw GenericErrorModel(customMessage:'A value should be provided');
```
Comment (StanevPrime) [inline packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart:68]: Using the `GenericErrorModel` makes the translation of the messaged shown to the user impossible

File: packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart [hunk 10]
```diff
@@ -1,7 +1,77 @@
...
+    }
+  }
+
+  void validateStreetOnType(String text) {
+    if (text.trim().isEmpty) {
+      throw GenericErrorModel(customMessage:'A value should be provided');
```
Comment (StanevPrime) [inline packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart:74]: Using the `GenericErrorModel` makes the translation of the messaged shown to the user impossible

File: packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart [hunk 11]
```diff
@@ -1,7 +1,77 @@
-import '../../../edit_address.dart';
+import '../../../widget_toolkit.dart';
```
Comment (StanevPrime) [inline packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart:1]: It's not good practice to import the entire package. Please import the specific files you need

**Review by StanevPrime — CHANGES_REQUESTED** [review]

File: packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart (hunk 3 above)
Comment (StanevPrime) [reply packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart]: Using the `GenericErrorModel` creates another issue. I have added a separate comment for it

File: packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart (hunk 4 above)
Comment (StanevPrime) [reply packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart]: Same as above

File: packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart [hunk 12]
```diff
@@ -1,7 +1,90 @@
...
+          fieldValue: text);
+    }
+    return text;
+  }
+
+  Future<String> validateStreetOnSubmit(String text) async {
+    if (text.trim().isEmpty) {
+      throw EditAddressErrorModel<String>(
+          error: 'A value should be provided',
+          translationKey: 'editAddressErrorModel',
```
Comment (StanevPrime) [inline packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart:67]: Please use an enum instead of a simple string where you will be able to add various scenarios if needed

**Review by StanevPrime — CHANGES_REQUESTED** [review]

File: packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart (hunk 12 above)
Comment (StanislavPrime) [reply packages/widget_toolkit/lib/src/lib_edit_address/services/edit_address_service.dart]: I changed the type of the `translationKey` to enum `EditAddressErrorType`

**Review by StanevPrime — APPROVED** [review]

File: packages/widget_toolkit/CHANGELOG.md [hunk 13]
```diff
@@ -1,9 +1,18 @@
 ## [0.0.1-dev6] - TBE
```
[the remark this answers is not in the document]
Comment (StanislavPrime) [reply packages/widget_toolkit/CHANGELOG.md]: TBE was deleted
